relation/src/gadgets/ecc/emulated/short_weierstrass.rs

```diff
@@ -0,0 +1,422 @@
+    /// - (x0 - x2) (y0 - y1) == (y0 + y2) (x0 - x1)
+    /// First check in both cases can be combined into the following:
+    /// inf0 == 1 || inf1 == 1 || inf2 == 0 || x0 != x1 || (y0 == y1 && y0 != 0)
+    /// For the rest equality checks,
+    ///   - Both LHS and RHS must be multiplied with an indicator variable
+    ///     (!inf0 && !inf1). So that if either p0 or p1 is infinity, those
+    ///     checks will trivially pass.
+    ///   - For the first case (point doubling), both LHS and RHS must be
+    ///     multiplied with an indicator variable (y0 != 0 && x0 == x1 && y1 ==
+    ///     y0). So that when y0 == 0 || x0 != x1 || y0 != y1, these checks will
+    ///     trivially pass.
+    ///   - For the second case, both LHS and RHS must be multiplied with (x0 -
+    ///     x1). So that when x0 == x1, these checks will trivially pass.
+    pub fn emulated_sw_ecc_add_gate<E: EmulationConfig<F>>(
+        &mut self,
+        p0: &EmulatedSWPointVariable<E>,
+        p1: &EmulatedSWPointVariable<E>,
+        p2: &EmulatedSWPointVariable<E>,
+        a: E,
+    ) -> Result<(), CircuitError> {
+        let eq_p1_p2 = self.is_emulated_sw_point_equal(p1, p2)?;
+        let eq_p0_p2 = self.is_emulated_sw_point_equal(p0, p2)?;
+        // Case 1: either p0 or p1 is infinity
+        self.enforce_equal(p0.2 .0, eq_p1_p2.0)?;
+        self.enforce_equal(p1.2 .0, eq_p0_p2.0)?;
+
+        // infinity_mark is 1 iff either p0 or p1 is infinity
+        let infinity_mark = self.logic_or(p0.2, p1.2)?;
+        // is 1 iff both p0 and p1 are not infinity
+        let non_infinity_mark = self.logic_neg(infinity_mark)?;
+
+        // Case 2: p2 is infinity, while p0 and p1 are not.
+        // inf0 == 1 || inf1 == 1 || inf2 == 0 || x0 != x1 || (y0 == y1 && y0 != 0)
+        let non_inf_p2 = self.logic_neg(p2.2)?;
+        let eq_x0_x1 = self.is_emulated_var_equal(&p0.0, &p1.0)?;
+        let neq_x0_x1 = self.logic_neg(eq_x0_x1)?;
+        let eq_y0_y1 = self.is_emulated_var_equal(&p0.1, &p1.1)?;
+        let is_y0_zero = self.is_emulated_var_zero(&p0.1)?;
+        let not_y0_zero = self.logic_neg(is_y0_zero)?;
+        let t = self.logic_and(eq_y0_y1, not_y0_zero)?;
+        let t = self.logic_or(neq_x0_x1, t)?;
+        let t = self.logic_or(non_inf_p2, t)?;
+        self.logic_or_gate(infinity_mark, t)?;
+
+        // Case 3: point doubling
+        // doubling mark is 1 iff x0 == x1 and y0 == y1
+        let doubling_mark = self.mul(eq_x0_x1.0, eq_y0_y1.0)?;
+        let doubling_coef = self.mul(doubling_mark, non_infinity_mark.0)?;
+        let doubling_coef = self.mul(doubling_coef, not_y0_zero.0)?;
+        // forcefully convert Variable into EmulatedVariable
+        // safe because it's boolean
+        let mut v = vec![self.zero(); E::NUM_LIMBS];
+        v[0] = doubling_coef;
+        let doubling_coef = EmulatedVariable::<E>(v, core::marker::PhantomData);
+
+        // first equality (x2 + 2 * x0) * (y0 + y0)^2 == (3 * x1^2 + a)^2
```

Comment:
Do we want to be consistent with code and use (x2+x0+x1) instead?

Reply:
Done in [cf9ff0e](https://github.com/EspressoSystems/jellyfish/pull/326/commits/cf9ff0ee56440e4767b8738978ab82b34c7afc32)